chronos/asyncsync.nim

```diff
@@ -610,3 +623,131 @@ template emitWait*[T](bus: AsyncEventBus, event: string,
   ## wait until all the subscribers/waiters will receive notification about
   ## event.
   emitWait(bus, event, data, getSrcLocation())
+
+proc `==`(a, b: EventQueueKey): bool {.borrow.}
+
+proc compact[T](ab: AsyncEventQueue[T]) =
+  if len(ab.readers) > 0:
+    let minOffset = ab.readers[0].offset
+    doAssert(minOffset >= ab.offset)
+    if minOffset > ab.offset:
+      let delta = minOffset - ab.offset
+      ab.queue.shrink(fromFirst = delta)
+      ab.offset += delta
+  else:
+    ab.queue.clear()
+
+proc getReaderIndex[T](ab: AsyncEventQueue[T], key: EventQueueKey): int =
+  for index, value in ab.readers.pairs():
+    if value.key == key:
+      return index
+  -1
+
+proc newAsyncEventQueue*[T](): AsyncEventQueue[T] =
+  ## Creates new ``AsyncEventBus``.
+  AsyncEventQueue[T](counter: 0'u64, queue: initDeque[T]())
+
+proc len*[T](ab: AsyncEventQueue[T]): int =
+  len(ab.queue)
+
+proc register*[T](ab: AsyncEventQueue[T]): EventQueueKey =
+  inc(ab.counter)
+  let reader = EventQueueReader(key: EventQueueKey(ab.counter),
+                                offset: ab.offset + len(ab.queue))
+  ab.readers.add(reader)
+  EventQueueKey(ab.counter)
+
+proc unregister*[T](ab: AsyncEventQueue[T], key: EventQueueKey) =
+  let index = ab.getReaderIndex(key)
+  if index >= 0:
+    let reader = ab.readers[index]
+    # Completing pending Future to avoid deadlock.
+    if not(isNil(reader.waiter)) and not(reader.waiter.finished()):
+      reader.waiter.complete()
+    ab.readers.delete(index)
+    ab.compact()
+
+proc emit*[T](ab: AsyncEventQueue[T], data: T) =
+  if len(ab.readers) > 0:
+    # We enqueue `data` only if there active reader present.
+    ab.queue.addLast(data)
+    for reader in ab.readers.items():
+      if not(isNil(reader.waiter)) and not(reader.waiter.finished()):
+        reader.waiter.complete()
+
+proc close*[T](ab: AsyncEventQueue[T]) =
+  for reader in ab.readers.items():
+    if not(isNil(reader.waiter)) and not(reader.waiter.finished()):
+      reader.waiter.complete()
+  # This could generate leak, shrink() is not yet implemented for sequences.
+  ab.readers.setLen(0)
+  ab.queue.clear()
+
+proc closeWait*[T](ab: AsyncEventQueue[T]): Future[void] =
+  var retFuture = newFuture[void]("AsyncEventQueue.closeWait()")
+  proc continuation(udata: pointer) {.gcsafe.} =
+    if not(retFuture.finished()):
+      retFuture.complete()
```

Comment:
Why? to raise `CancelledError` from `closeWait()` procedure?
The idea here is to ensure that all event consumers will receive notification about that `AsyncEventQueue` is get closed, and because all this callbacks get schedule inside `ab.close()` call, it means that `continuation` procedure will be scheduled after all the consumer notification callbacks will be already processed.